### PR TITLE
Monitoring

### PR DESCRIPTION
**Pull request by NickCraver**
Initial pass at a monitoring blog post (and a few misc styling commits needed for it)

### CODE REVIEW

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+It's now breaking every page.
+A health check route running some code wouldn't trigger, but just the act of loading the master view ensures a huge number of dependencies are evaluated and working for the check.
+
+If you're curious, that's not a hypothetical.
+You know that little dot on the review queue that indicates a lot of items currently in queue?
+Yeah...fun Tuesday.
+
+We also have health checks inside libraries.
+The simplest manifestation of this is a [heartbeat](https://en.wikipedia.org/wiki/Heartbeat_(computing)).
+This is something that for example [StackExchage.Redis](https://github.com/StackExchange/StackExchange.Redis) uses to routinely check if the socket connection to Redis is active.
```
Comment (m0sa) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
This is something that for example [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis) uses to routinely check if the socket connection to Redis is active.
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+Is the certificate valid? (We couldn't connect if it isn't.)
+Does the firewall allow the rule?
+
+By having something continually checking this and feeding into alerts when it fails, we can quickly identify issues, especially those from invalid config changes to the infrastructure.
+We can also readily test new configurations or infrastructure, firewall rules, etc. before user load is applied.
+For more details, see [the GitHub README][GH.httpUnit].
+
+#### Health Checks: Fastly
+
+If we zoom out fom the data center, we need to see what's hitting us.
```
Comment (m0sa) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
If we zoom out from the data center, we need to see what's hitting us.
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+- Name: `dotnet.memory.gc_collections`
+- Time: `2018-01-01 18:30:00` (Of course it's UTC, we're not barbarians.)
+- Value: `129,389,139`
+- Tags: Server: `NY-WEB01`, Application: `StackExchange-Network`
+
+The value in an entry can also take a few forms, but the general case is counters.
+Counters report an ever-increasing value (often reset to `0` on restarts and such though).
+By taking the difference in value over time, you can find out the value delta in that window.
+For example, if we had 129,389,039 ten minutes before, we know that process on that server in those ten minutes ran 100 Gen 0 garbage collection passes.
+Another case is just reporting an absolute point-in-time value, for example "This GPU is currenly 87°".
```
Comment (m0sa) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
Another case is just reporting an absolute point-in-time value, for example "This GPU is currently 87°".
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+
+...hmmm, good questions!
+These are the things that matter to the user experience.
+Our question pages render in 18ms.
+I think that's awesome.
+And I might even be biased.
+...but it also doesn't mean crap to a user if it takes forever to get to them.
+
+So, what can we do?
+Years back, I threw together a client timings pipline when the pieces we needed first became available in browsers.
```
Comment (m0sa) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
Years back, I threw together a client timings pipeline when the pieces we needed first became available in browsers.
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 5]
```diff
@@ -0,0 +1,852 @@
...
+
+#### Opserver: FAQs
+
+I get the same questions about Opserver routinely, so let's knock a few of them out:
+- Opserver does **not** require a data store of any kind for itself (it's all config and in-memory state).
+  - This may happen in the future to enhance functionality, but there are no plans to require anything.
+- Only the dashboard tab and per-node view is powered by Bosun, Orion, or WMI - all other screens like SQL, Elastic, Redis, etc. have no dependency...Opserver monitors these directly.
+- Authentication is both global and per-tab pluggable (who can view and who's an admin are separate), but built-in configuration is via groups and Active Directory is included.
+  - On admin vs. viewer: A viewer gets a read-only view. For example, HAProxy controls wouldn't be shown.
+- All tabs are not requied--each is independent and only appears if configured.
```
Comment (m0sa) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
- All tabs are not required--each is independent and only appears if configured.
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 6]
```diff
@@ -0,0 +1,852 @@
...
+#### Bosun Next Steps
+
+Bosun has largely been in maintenance mode due to a lack of resources and other priorities.
+We haven't done as much as we'd like because we need to have the discussion on the best path forward.
+Have other tools caught up and filled the gaps that caused us to build it in the first place?
+Has SQL 2017 or 2019 already improved the queries we had issues with lowering the bar greatly?
+We need to take some time and look at the landscape and evaluate what we want to do.
+This is something we want to get into during Q1 2019.
+
+We know of some things we'd like to do, such as impoving the alert editing experience and some other UI areas.
```
Comment (m0sa) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
We know of some things we'd like to do, such as improving the alert editing experience and some other UI areas.
```

**Review by m0sa — CHANGES_REQUESTED** [review]
fixing typos

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 7]
```diff
@@ -0,0 +1,853 @@
...
+Well, it's a trade-off.
+Have you ever run a program with tons of console output?
+Run the same program without it? Goes faster, doesn't it?
+Logging has a few costs.
+First, you often need to allocate strings for the logging itself.
+That's memory and garbage collection (for .NET and some other platforms).
+When you're logging *somewhere*, that usually that means disk space.
+If we're traversing a network (and to some degree locally), it also means bandwidth and latency.
+
+...and I was just kidding about email and [GDPR](https://en.wikipedia.org/wiki/General_Data_Protection_Regulation)...**all of those are GDPR concerns**. Keep retention and compliance in mind when logging anything. It's another cost to consider.
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
...and I was just kidding about [GDPR](https://en.wikipedia.org/wiki/General_Data_Protection_Regulation) only being a concern for email...**GDPR is a concern for all of the above**. Keep retention and compliance in mind when logging anything. It's another cost to consider.
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 8]
```diff
@@ -0,0 +1,853 @@
...
+For the record, [Jarrod Dixon](https://twitter.com/jarrod_dixon) first suggested and started the HTTP logging about 8 years ago and we all told him he was an insane lunatic and it was a tremendous waste of resources.
+Please no one tell him he was totally right.
+A new per-month storage format will be coming soon, but that's another story.
+
+In those requests, we use profiling [that we'll talk about shortly](#miniprofiler) and send headers to HAProxy with certain numbers.
+HAProxy captures and strips those headers into the syslog row we forward for processing into SQL.
+Those headers include:
+
+- ASP.NET Overall Milliseconds (encompasses those below)
+- SQL Count & Milliseconds
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
- SQL Count (queries) & Milliseconds
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 9]
```diff
@@ -0,0 +1,853 @@
...
+- Redis Count & Milliseconds
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
- Redis Count (hits) & Milliseconds
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 10]
```diff
@@ -0,0 +1,853 @@
...
+- HTTP Count & Milliseconds
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
- HTTP Count (requests sent) & Milliseconds
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 11]
```diff
@@ -0,0 +1,853 @@
...
+- Tag Engine Count & Milliseconds
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
- Tag Engine Count (queries) & Milliseconds
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 12]
```diff
@@ -0,0 +1,853 @@
...
+- Elasticsearch Count & Milliseconds
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
- Elasticsearch Count (hits) & Milliseconds
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 13]
```diff
@@ -0,0 +1,853 @@
...
+Does the firewall allow the rule?
+
+By having something continually checking this and feeding into alerts when it fails, we can quickly identify issues, especially those from invalid config changes to the infrastructure.
+We can also readily test new configurations or infrastructure, firewall rules, etc. before user load is applied.
+For more details, see [the GitHub README][GH.httpUnit].
+
+#### Health Checks: Fastly
+
+If we zoom out from the data center, we need to see what's hitting us.
+That's [Fastly](https://www.fastly.com/).
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
That's usually our CDN & proxy: [Fastly](https://www.fastly.com/).
```

File: blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown [hunk 14]
```diff
@@ -0,0 +1,853 @@
...
+Fastly has a concept of [services](https://docs.fastly.com/guides/basic-setup/working-with-services), which are akin to HAProxy backends when you think about it like a load balancer.
+Fastly also has [health checks](https://docs.fastly.com/guides/basic-configuration/working-with-health-checks) built in.
+In each of our data centers, we have two sets of ISPs coming in for redundancy.
+We can configure things in Fastly to optimize uptime here.
+
+Let's say our NY data center is primary at the moment, and CO is our backup (our default state).
```
Comment (NickCraver) [inline blog/_posts/2018-11-29-stack-overflow-how-we-do-monitoring.markdown]:
```suggestion
Let's say our NY data center is primary at the moment, and CO is our backup.
```

**Review by NickCraver — COMMENTED** [review]
Notes from Tom, thanks!

**Comment by NickCraver** [comment]
Go time! Merging in.

**Pull request merged**